service-base/src/main/java/org/eclipse/hono/service/AbstractApplication.java

```diff
@@ -95,212 +58,45 @@ public final void addServiceFactories(final Set<ObjectFactory<? extends Abstract
-     * 
-     * @param maxWaitTime The maximum time to wait for the server to shut down (in seconds).
-     * @param shutdownHandler The handler to invoke with the result of the shutdown attempt.
-     */
-    public final void shutdown(final long maxWaitTime, final Handler<Boolean> shutdownHandler) {
-
-        try {
-            log.info("shutting down application...");
-
-            preShutdown();
-            final CountDownLatch latch = new CountDownLatch(1);
-
-            stopHealthCheckServer().setHandler(result -> {
-
-                if (vertx != null) {
-                    log.info("closing vert.x instance ...");
-                    vertx.close(r -> {
-                        if (r.failed()) {
-                            log.error("could not close vert.x instance", r.cause());
-                        }
-                        latch.countDown();
-                    });
-                } else {
-                    latch.countDown(); // nothing to wait for
-                }
-            });
-
-            if (latch.await(maxWaitTime, TimeUnit.SECONDS)) {
-                log.info("application has been shut down successfully");
-                shutdownHandler.handle(Boolean.TRUE);
-            } else {
-                log.error("shut down timed out, aborting...");
-                shutdownHandler.handle(Boolean.FALSE);
-            }
-        } catch (final InterruptedException e) {
-            log.error("application shut down has been interrupted, aborting...");
-            Thread.currentThread().interrupt();
-            shutdownHandler.handle(Boolean.FALSE);
-        }
-    }
-
-    private Future<Void> stopHealthCheckServer() {
-        return healthCheckServer.stop();
-    }
-
-    /**
-     * Invoked after the service instances have been deployed successfully.
-     * <p>
-     * May be overridden to provide additional startup logic, e.g. deploying
-     * additional verticles.
-     * <p>
-     * This default implementation simply returns a succeeded future.
-     * 
-     * @return A future indicating success. Application start-up fails if the
-     *         returned future fails.
-     */
-    protected Future<?> postRegisterServiceVerticles() {
-        return Future.succeededFuture();
-    }
-
-    /**
-     * Invoked before application shutdown is initiated.
-     * <p>
-     * May be overridden to provide additional shutdown handling, e.g.
-     * releasing resources before the vert.x instance is closed.
-     */
-    protected void preShutdown() {
-        // empty
+    @Override
+    protected Future<?> deployVerticles() {
+        // call into super ...
+        return super.deployVerticles()
+                // ... then deploy service verticles
+                .compose(s -> deployServiceVerticles(getConfig().getMaxInstances()))
+                // ... then deploy the required verticles
+                .compose(s -> deployRequiredVerticles(getConfig().getMaxInstances()));
```

Comment:
shouldn't this be the other way around, i.e. first deploy the required verticles, then the service verticles?